tests/vectors_sig.c

```diff
@@ -165,109 +212,361 @@ OQS_STATUS sig_vector(const char *method_name,
+		ret = EXIT_SUCCESS;
+	} else {
+		ret = EXIT_FAILURE;
+		fprintf(stderr, "[vectors_sig] %s ERROR: public key or private key doesn't match!\n", method_name);
+	}
+	goto cleanup;
 
-	hexStringToByteArray(prng_output_stream, prng_output_stream_bytes);
-	hexStringToByteArray(sig_msg, sig_msg_bytes);
-	hexStringToByteArray(sig_sk, sig_sk_bytes);
-	hexStringToByteArray(verif_sig, verif_sig_bytes);
-	hexStringToByteArray(verif_pk, verif_pk_bytes);
-	hexStringToByteArray(verif_msg, verif_msg_bytes);
+err:
+	ret = EXIT_FAILURE;
+	goto cleanup;
 
-	rc = sig_vector(alg_name, prng_output_stream_bytes, sig_msg_bytes, sig_msg_len, sig_sk_bytes, verif_sig_bytes, verif_pk_bytes, verif_msg_bytes, verif_msg_len);
+algo_not_enabled:
+	ret = EXIT_SUCCESS;
+
+cleanup:
+	OQS_MEM_insecure_free(signature);
+	OQS_SIG_free(sig);
+	return ret;
+}
+#endif
+
+
+
+int main(int argc, char **argv) {
+	OQS_STATUS rc = OQS_SUCCESS;
+	bool valid_args = true;
+
+	OQS_init();
+
+	size_t msgLen;
+
+	char *prng_output_stream = NULL;
+	char *kg_pk;
+	char *kg_sk;
+
+	char *sigGen_sk;
+	char *sigGen_msg;
+	char *sigGen_sig;
+
+	char *sigVer_pk;
+	char *sigVer_msg;
+	char *sigVer_sig;
+
+	uint8_t *prng_output_stream_bytes = NULL;
+	uint8_t *kg_pk_bytes = NULL;
+	uint8_t *kg_sk_bytes = NULL;
+
+	uint8_t *sigGen_sk_bytes = NULL;
+	uint8_t *sigGen_msg_bytes = NULL;
+	uint8_t *sigGen_sig_bytes = NULL;
+
+
+	uint8_t *sigVer_pk_bytes = NULL;
+	uint8_t *sigVer_msg_bytes = NULL;
+	uint8_t *sigVer_sig_bytes = NULL;
+
+	OQS_SIG *sig = NULL;
+
+	if (argc < 3) {
+		valid_args = false;
+		goto err;
+	}
+
+	char *alg_name = argv[1];
+	char *test_name = argv[2];
+
+	sig = OQS_SIG_new(alg_name);
+
+	if (sig == NULL) {
+		printf("[vectors_sig] %s was not enabled at compile-time.\n", alg_name);
+		goto err;
+	}
+
+	if (!strcmp(test_name, "keyGen")) {
+		if (argc != 6) {
+			valid_args = false;
+			goto err;
+		}
+		prng_output_stream = argv[3]; // d || z
+		kg_pk = argv[4];
+		kg_sk = argv[5];
+
+		if (strlen(prng_output_stream) % 2 != 0 ||
+		        strlen(kg_pk) != 2 * sig->length_public_key ||
+		        strlen(kg_sk) != 2 * sig->length_secret_key) {
+			printf("lengths bad\n");
+			goto err;
+		}
+
+		prng_output_stream_bytes = OQS_MEM_malloc(strlen(prng_output_stream) / 2);
+		kg_pk_bytes = OQS_MEM_malloc(sig->length_public_key);
+		kg_sk_bytes = OQS_MEM_malloc(sig->length_secret_key);
+
+		if ((prng_output_stream_bytes == NULL) || (kg_pk_bytes == NULL) || (kg_sk_bytes == NULL)) {
+			fprintf(stderr, "[vectors_sig] ERROR: OQS_MEM_malloc failed!\n");
+			goto err;
+		}
+
+		hexStringToByteArray(prng_output_stream, prng_output_stream_bytes);
+		hexStringToByteArray(kg_pk, kg_pk_bytes);
+		hexStringToByteArray(kg_sk, kg_sk_bytes);
+
+
+		rc = sig_kg_vector(alg_name, prng_output_stream_bytes, kg_pk_bytes, kg_sk_bytes);
+
+	} else if (!strcmp(test_name, "sigGen_det") || !strcmp(test_name, "sigGen_rnd")) {
+		int randomized = !strcmp(test_name, "sigGen_rnd");
```

Comment:
I think this is a bitwise NOT, which is OK for the truthiness comparison in the `else if` clause but will cause problems if we treat it as an arithmetic value. Probably should do something like this instead:
```suggestion
		int randomized = (strcmp(test_name, "sigGen_rnd") != 0);
```

Reply:
Wouldn't `~` be the bitwise NOT operator, while `!` is the logical NOT operator, which always returns either 0 or 1?

Reply:
Of course, you are correct, disregard.